### ISSUE TITLE
problem running ExecuteNetwork test

### ISSUE THREAD

**Issue by liviolima80** [ARM-software/armnn#7]
Hi,
I'm trying to test ExecuteNetwork example on LeNet example in Caffe tutorial

http://nbviewer.jupyter.org/github/BVLC/caffe/blob/master/examples/01-learning-lenet.ipynb

After training process I have both binary weights file and the prototxt deploy model attached
[new_shape_deploy.zip](https://github.com/ARM-software/armnn/files/1884097/new_shape_deploy.zip)

If I run ExecuteNetwork passing binary file I have the following error
 
Armnn Error: Unsupported layer type 'Data'

since only "Input" type layer is supported. 

If I run ExecuteNetwork passing text file I have the following error

Armnn Error: Expected data blob at index 0 in layer conv1 not found

Can someone help me to solve the issue?

**Comment by MatthewARM** [comment]
Hi @liviolima80 the Caffe parser doesn't support the 'Data' layer: the supported layers are listed in CaffeSupport.md in the armnnCaffeParser source directory. However I am aware that we have a working test using Lenet. I'll find out which version of the model we're using and post back here.

**Comment by liviolima80** [comment]
Hi @MatthewARM , any news?

**Comment by TelmoARM** [comment]
Hi @liviolima80, I've been looking into the issues you reported and I will have some news for you today

**Comment by TelmoARM** [comment]
Hi,

Apologies for the delay, I had to clarify a few things before I could provide a clear answer. I was able to reproduce the errors you reported. 

1 - Passing a caffe-text input: 

The error is because the .protoxt file doesn't have the weights included. In Tensorflow you can merge the network weights with the model definition in the .protoxt but Caffe doesn't really support this, it was an oversight on our side and we will be clarifying the expected behaviour on our next release. The caffe-binary input for Lenet is working, see below.

1.1 Armnn Error: Unsupported layer type 'Data'

It's case sensitive. Change the input layer to '-i data'	

2 - Alexnet caffe-binary input not working. 
   
I was able to reproduce this issue as well. We have an Alexnet test that runs fine outside the ExecuteNetwork app, so we must be doing something wrong in the app. I'll open a bug to investigate this further and we'll let you know of the outcome.

**Comment by TelmoARM** [comment]
Hi again @liviolima80,

It appears that my 1.1 answer above is incorrect. That error is caused when you use a caffe .prototxt with the old syntax (which Armnn doesn't support). That can be easily fixed with something like:
import caffe
net = caffe.Net('old.prototxt', '.caffemodel', caffe.TEST)
net.save('new.caffemodel') 

Let us know if that works

**Comment by liviolima80** [comment]
Hi @TelmoARM ,
first of all thanks for your reply, i'll try the python fix you suggest. The problem is that the loading of a caffe model in python does not work in windows, i have to do in linux

In the meantime, can you provide me the pyton code you use to build the model for train in caffe in order to create a model that works with armnn? I usually use the following 

n = caffe.NetSpec()

n.data, n.label = L.Data(batch_size=batch_size, backend=P.Data.LMDB, source=lmdb, transform_param=dict(scale=1./255), ntop=2)
    
n.conv1 = L.Convolution(n.data, kernel_size=3, num_output=32, weight_filler=dict(type='xavier'))
n.relu1 = L.ReLU(n.conv1, in_place=True)
n.conv2 = L.Convolution(n.relu1, kernel_size=3, num_output=32, weight_filler=dict(type='xavier'))
n.relu2 = L.ReLU(n.conv2, in_place=True)
n.pool1 = L.Pooling(n.relu2, kernel_size=2, stride=2, pool=P.Pooling.MAX) 

...... 

Thanks

**Comment by liviolima80** [comment]
Hi @TelmoARM ,
with your fix now the model works! Now I have another problem that the network output is wrong but probably I load the data to input tensor in wrong way. I need to investigate about this

Thanks you

**Comment by TelmoARM** [comment]
Hi @liviolima80 ,

Good to hear the fix worked. 

I've also found the problem with Alexnet. The Alexnet deploy.prototxt has got the batch size set to 10 but Armnn uses batch size in the model and there's no way to override it, so we need to generate a new model with the batch size set to 1. To do this:

Open the Alexnet deploy.prototxt and change 'input_param { shape: { dim: 10 dim: 3 dim: 227 dim: 227 } }' to 'input_param { shape: { dim: 1 dim: 3 dim: 227 dim: 227 } }' and save it as Alexnet1.prototxt.

Run the following python script
--------------
import caffe

net = caffe.Net('deploy.prototxt', '/home/bvlc_alexnet.caffemodel', caffe.TEST)
new_net = caffe.Net('Alexnet1.prototxt', '/home/bvlc_alexnet.caffemodel', caffe.TEST)
new_net.save('bvlc_batchsize_1.caffemodel') 
----------------

The new bvlc_batchsize_1.caffemodel runs through Armnn.

**Comment by MatthewARM** [comment]
Closing this ticket as I believe both issues have been solved.

**Comment by GarryLau** [comment]
How to solve the problem ？Armnn Error: Couldn't find requested output layer 'prob' in graph. I used the trained caffemodel directly. Should I convert it OR use both develop.prototxt and trained caffemodel？

**Comment by GarryLau** [comment]
@TelmoARM I have solved my problem. I have wrote a [blog](https://blog.csdn.net/liugan528/article/details/80430760) (of course in Chinese),  and I put my codes in [github](https://github.com/GarryLau/armnn_caffe). 
I have put my code on Ambarella platform， and running successfully.
I think your document are too few, you should tell us we must convert the trained caffemodel to the inference caffemodel, and something important we need to attend to about cross compile. (I also wrote a blog about how to configure the develop environment about armnn sdk, compute library, protobuf and so on.)
